src/db/scan.rs

```diff
@@ -0,0 +1,502 @@
+use crate::prolly;
+
+#[allow(dead_code)]
+pub struct ScanKey<'a> {
+    value: &'a [u8],
+    exclusive: bool,
+}
+
+#[allow(dead_code)]
+pub struct ScanBound<'a> {
+    // TODO: Make these two fields exclusive?
+    key: Option<ScanKey<'a>>,
+    index: Option<u64>,
+}
+
+#[allow(dead_code)]
+pub struct ScanOptions<'a> {
+    // TODO: Make these two fields exclusive?
+    prefix: Option<&'a [u8]>,
+    start: Option<ScanBound<'a>>,
+    limit: Option<u64>,
+}
+
+#[allow(dead_code)]
+pub fn scan<'a>(
+    map: &'a prolly::Map,
+    opts: ScanOptions<'a>,
+) -> impl Iterator<Item = prolly::Entry<'a>> {
+    let mut it = map.iter().peekable();
+    let mut prefix: &[u8] = &[];
+    let mut from_key: &[u8] = &[];
+    let mut from_index = 0u64;
+    let mut exclusive = false;
+
+    if let Some(p) = opts.prefix {
+        from_key = p;
+        prefix = p;
+    }
+    if let Some(start) = opts.start {
+        if let Some(key) = start.key {
+            if key.value > from_key {
+                from_key = key.value;
+                exclusive = key.exclusive;
+            }
+        }
+        if let Some(index) = start.index {
+            from_index = index;
+        }
+    }
+
+    let key_met = |key: &[u8]| {
+        if exclusive {
+            key > from_key
+        } else {
+            key >= from_key
+        }
+    };
+
+    let mut index = 0;
+    while it.peek().is_some() {
```

Comment:
can peek block, eg going to disk?

Reply:
Not currently. The entire read API of prolly::Map is synchronous currently because the map is held in memory. Once prolly trees are implemented all the methods will become asynchronous.